src/ray/raylet/pubsub/pubsub_client.h

```diff
@@ -0,0 +1,162 @@
+                                 const ObjectID &object_id) = 0;
+
+  virtual ~PubsubClientInterface() {}
+};
+
+/// The pubsub client implementation.
+class PubsubClient : public PubsubClientInterface {
+ public:
+  explicit PubsubClient(const NodeID self_node_id, const std::string self_node_address,
+                        const int self_node_port,
+                        rpc::CoreWorkerClientPool &owner_client_pool)
+      : self_node_id_(self_node_id),
+        self_node_address_(self_node_address),
+        self_node_port_(self_node_port),
+        owner_client_pool_(owner_client_pool) {}
+  ~PubsubClient() = default;
+
+  void SubcribeObject(const rpc::Address &owner_address, const ObjectID &object_id,
+                      SubscriptionCallback subscription_callback,
+                      SubscriptionFailureCallback subscription_failure_callback) override;
+
+  bool UnsubscribeObject(const rpc::Address &owner_address,
+                         const ObjectID &object_id) override;
+
+ private:
+  /// Encapsulates the subscription information such as subscribed object ids ->
+  /// callbacks.
+  /// TODO(sang): Use a channel abstraction instead of owner address once OBOD uses the
+  /// pubsub.
+  struct SubscriptionInfo {
+    SubscriptionInfo(const rpc::Address &pubsub_server_address)
+        : pubsub_server_address_(pubsub_server_address) {}
+
+    /// Address of the pubsub server.
+    const rpc::Address pubsub_server_address_;
+    // Object ID -> subscription_callback
+    absl::flat_hash_map<const ObjectID,
+                        std::pair<SubscriptionCallback, SubscriptionFailureCallback>>
+        subscription_callback_map_;
+  };
+
+  /// Create a long polling connection to the owner for receiving the published messages.
+  /// TODO(sang): Currently, we assume that unregistered objects will never be published
+  /// from the pubsub server. We may want to loose the restriction once OBOD is supported
+  /// by this function.
+  /// \param owner_address The address of the owner that publishes
+  /// objects.
+  /// \param subscriber_address The address of the subscriber.
+  void MakeLongPollingPubsubConnection(const rpc::Address &owner_address,
+                                       const rpc::Address &subscriber_address);
+
+  /// Private method to handle long polling responses. Long polling responses contain the
+  /// published messages.
+  void HandleLongPollingResponse(const rpc::Address &owner_address,
+                                 const rpc::Address &subscriber_address, Status &status,
+                                 const rpc::PubsubLongPollingReply &reply);
+
+  /// Private method to handle pubsub server failures.
+  void HandleCoordinatorFailure(const rpc::Address &owner_address);
+
+  /// Returns a subscription callback; The first entry indicates if the subscription
+  /// callback exists. The second entry is the callback itself.
+  inline std::pair<bool, SubscriptionCallback> GetSubscriptionCallback(
```

Comment:
Couldn't this just be a `SubscriptionCallback` return type, since a returned `nullptr` would indicate that the callback doesn't exist?

Reply:
I generally don't prefer to return nullptr. I can use optional instead if you prefer that.

Reply:
Yeah I think an `absl::optional<SubscriptionCallback>` would be the best out of the three options.